### PR TITLE
Meta-internal folly integration with F14FastMap

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: Especially after updating to C++17, I don't see a compelling case for
*requiring* any folly components in RocksDB. I was able to purge the existing
hard dependencies, and it can be quite difficult to strip out non-trivial components
from folly for use in RocksDB. (The prospect of doing that on F14 has changed
my mind on the best approach here.)

But this change creates an optional integration where we can plug in
components from folly at compile time, starting here with F14FastMap to replace
std::unordered_map when possible (probably no public APIs for example). I have
replaced the biggest CPU users of std::unordered_map with compile-time
pluggable UnorderedMap which will use F14FastMap when USE_FOLLY is set.
USE_FOLLY is always set in the Meta-internal buck build, and a simulation of
that is in the Makefile for public CI testing. A full folly build is not needed, but
checking out the full folly repo is much simpler for getting the dependency,
and anything else we might want to optionally integrate in the future.

Some picky details:
* I don't think the distributed mutex stuff is actually used, so it was easy to remove.
* I implemented an alternative to `folly::constexpr_log2` (which is much easier
in C++17 than C++11) so that I could pull out the hard dependencies on
`ConstexprMath.h`
* I had to add noexcept move constructors/operators to some types to make
F14's complainUnlessNothrowMoveAndDestroy check happy, and I added a
macro to make that easier in some common cases.
* Updated Meta-internal buck build to use folly F14Map (always)

No updates to HISTORY.md nor INSTALL.md as this is not (yet?) considered a
production integration for open source users.

Test Plan: CircleCI tests updated so that a couple of them use folly.

Most internal unit & stress/crash tests updated to use Meta-internal latest folly.
(Note: they should probably use buck but they currently use Makefile.)

Example performance improvement: when filter partitions are pinned in cache,
they are tracked by PartitionedFilterBlockReader::filter_map_ and we can build
a test that exercises that heavily. Build DB with

```
TEST_TMPDIR=/dev/shm/rocksdb ./db_bench -benchmarks=fillrandom -num=10000000 -disable_wal=1 -write_buffer_size=30000000 -bloom_bits=16 -compaction_style=2 -fifo_compaction_max_table_files_size_mb=10000 -fifo_compaction_allow_compaction=0 -partition_index_and_filters
```

and test with (simultaneous runs with & without folly, ~20 times each to see
convergence)

```
TEST_TMPDIR=/dev/shm/rocksdb ./db_bench_folly -readonly -use_existing_db -benchmarks=readrandom -num=10000000 -bloom_bits=16 -compaction_style=2 -fifo_compaction_max_table_files_size_mb=10000 -fifo_compaction_allow_compaction=0 -partition_index_and_filters -duration=40 -pin_l0_filter_and_index_blocks_in_cache
```

Average ops/s no folly: 26229.2
Average ops/s with folly: 26853.3 (+2.4%)

### CODE REVIEW

**Comment by mrambacher** [comment]
I agree we should use folly where it makes sense.  However, I question why you want it in the source tree.  Why not detect that folly is installed on the machine (comparable to Snappy or other third-party code) and use it if available and instructed to?  Over time I could see more and more classes that could potentially have Folly-backed alternatives and I do not know that you want each to keep managing those in third-party.

**Comment by mrambacher** [comment]
> > However, I question why you want it in the source tree.
> 
> To be clear, with this change it's temporarily in the source tree, not checked in or even formally submoduled. And if we build something under our build command, it should be under our source tree. (Code navigation, etc.)

I am not a fan of copying code from another repo into the RocksDB one.  I think using a "submodule" that can optionally be added in would work.  Temporary added code tends to become "permanent" as priorities change...

> > Why not detect that folly is installed on the machine (comparable to Snappy or other third-party code) and use it if available and instructed to?
> 
> I believe folly follows a much more agile development cycle than the compression libraries, which means we have a lot more compatibility problems if we try to be compatible with many different versions of folly. The best way to pin each RocksDB revision to a particular folly revision is to just check it out and build the parts we need.

I have seen other projects solve this by checking out a specific version of folly (or other tools).  Velox and other related projects target a specific version of folly.  That would be preferable to me.  

It would also be great if folly was not so monolithic and it was easier to pick and choose pieces, but that is a dream for another project...

**Comment by pdillinger** [comment]
> I am not a fan of copying code from another repo into the RocksDB one. I think using a "submodule" that can optionally be added in would work. Temporary added code tends to become "permanent" as priorities change...

I don't see what the problem of checking out code into a git-ignored directory is. It's as likely to become "permanent" as compiled binaries (which has happened btw: 4abf94eeb4e894b979b69a7aef72c1bd619d8762)

Some googling suggests there isn't a built-in or easy notion of **optional** submodules.

> I have seen other projects solve this by checking out a specific version of folly (or other tools). Velox and other related projects target a specific version of folly. That would be preferable to me.

Like this?
https://github.com/facebook/rocksdb/pull/9546/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R2404

> It would also be great if folly was not so monolithic and it was easier to pick and choose pieces, but that is a dream for another project...

In the buck build world it is not monolithic, but it's unlikely we're going to move open source users over to buck.

File: src.mk [hunk 1]
```diff
@@ -364,12 +364,11 @@ TEST_LIB_SOURCES =                                              \
...
-  third-party/folly/folly/detail/Futex.cpp                                     \
-  third-party/folly/folly/synchronization/AtomicNotification.cpp               \
-  third-party/folly/folly/synchronization/DistributedMutex.cpp                 \
-  third-party/folly/folly/synchronization/ParkingLot.cpp                       \
-  third-party/folly/folly/synchronization/WaitOptions.cpp                      \
+FOLLY_SOURCES =                                                 \
+  $(FOLLY_DIR)/folly/container/detail/F14Table.cpp              \
+  $(FOLLY_DIR)/folly/lang/SafeAssert.cpp                        \
+  $(FOLLY_DIR)/folly/lang/ToAscii.cpp                           \
+  $(FOLLY_DIR)/folly/ScopeGuard.cpp                             \
```
Comment (ajkr) [inline src.mk:371]: There's four things here but one in buck ("//folly/container:f14_hash") - does that mean the one buck target brings with it the other three sources?

**Review by ajkr — APPROVED** [review]
Awesome! Look forward to try using folly.

File: src.mk (hunk 1 above)
Comment (pdillinger) [reply src.mk:371]: Yes, through deps

File: src.mk [hunk 2]
```diff
@@ -364,12 +364,11 @@ TEST_LIB_SOURCES =                                              \
   test_util/testutil.cc                                         \
   utilities/cassandra/test_utils.cc                             \
 
-FOLLY_SOURCES = \
-  third-party/folly/folly/detail/Futex.cpp                                     \
-  third-party/folly/folly/synchronization/AtomicNotification.cpp               \
-  third-party/folly/folly/synchronization/DistributedMutex.cpp                 \
-  third-party/folly/folly/synchronization/ParkingLot.cpp                       \
-  third-party/folly/folly/synchronization/WaitOptions.cpp                      \
+FOLLY_SOURCES =                                                 \
```
Comment (anand1976) [inline src.mk:367]: IIUC, this is required only for the test-only simulation right? if we use a folly feature for Meta internal purposes, and it's guarded by something like ```#if defined(USE_FOLLY && FBCODE_ONLY)```, we'd just specify a dependency in the TARGETS file.

Comment (pdillinger) [reply src.mk:367]: That's right. The issue that is blocking this PR now is that our internal crash tests use our own Makefile builds (which is dubious for maximizing similarity to internal buck builds) and it's proving tricky to get this Makefile build working with internal folly. And that is pushing me toward moving us toward using buck builds for internal crash tests, and that's a bit of work.

**Pull request closed without merging**